File: public/docs/_examples/server-communication/ts/app/wiki/wiki-smart.component.ts
/* tslint:disable: member-ordering forin */
// #docplaster
// #docregion
import { Component, OnInit }        from '@angular/core';
import { Observable }       from 'rxjs/Observable';
// #docregion import-subject
import { Subject }          from 'rxjs/Subject';
// #enddocregion import-subject

import { WikipediaService } from './wikipedia.service';

@Component({
  moduleId: module.id,
  selector: 'my-wiki-smart',
  templateUrl: './wiki.component.html',
  providers: [ WikipediaService ]
})
export class WikiSmartComponent implements OnInit {
  title   = 'Smarter Wikipedia Demo';
  fetches = 'Fetches when typing stops';
  items: Observable<string[]>;

  // #docregion subject
  private searchTermStream = new Subject<string>();
  search(term: string) { this.searchTermStream.next(term); }
  // #enddocregion subject

  constructor (private wikipediaService: WikipediaService) {}

  ngOnInit() {
    // #docregion observable-operators
    this.items = this.searchTermStream
      .debounceTime(300)
      .distinctUntilChanged()
      .switchMap((term: string) => this.wikipediaService.search(term));
    // #enddocregion observable-operators
  }
}

File: public/docs/_examples/server-communication/dart/lib/wiki/wiki_smart_component.dart
// #docregion
import 'package:angular2/core.dart';
import 'package:stream_transformers/stream_transformers.dart';

import 'wikipedia_service.dart';

@Component(
    selector: 'my-wiki-smart',
    template: '''
      <h1>Smarter Wikipedia Demo</h1>
      <p><i>Fetches when typing stops</i></p>

      <input #term (keyup)="search(term.value)"/>
      <ul>
        <li *ngFor="let item of items">{{item}}</li>
      </ul>
    ''',
    providers: const [WikipediaService])
class WikiSmartComponent {
  final WikipediaService _wikipediaService;
  List items = [];

  WikiSmartComponent(this._wikipediaService) {
    _searchTermStream
        .transform(new Debounce(new Duration(milliseconds: 300)))
        .distinct()
        .transform(new FlatMapLatest(
            (term) => _wikipediaService.search(term).asStream()))
        .forEach((data) {
      items = data;
    });
  }

  final EventEmitter _searchTermStream = new EventEmitter();

  void search(String term) => _searchTermStream.add(term);
}
